src/github.com/matrix-org/dendrite/roomserver/input/state.go

```diff
@@ -0,0 +1,258 @@
+package input
+
+import (
+	"fmt"
+	"github.com/matrix-org/dendrite/roomserver/types"
+	"github.com/matrix-org/gomatrixserverlib"
+	"sort"
+)
+
+func calculateAndStoreState(
+	db RoomEventDatabase, event gomatrixserverlib.Event, roomNID types.RoomNID, stateEventIDs []string,
+) (types.StateSnapshotNID, error) {
+	if stateEventIDs != nil {
+		// 1) We've been told what the state at the event is.
```

Comment:
They are left over from when I was enumerating the cases I wanted to handle. In a later PR I'd like to go back and hook up metrics for each of the cases.